elastica/experimental/connection_contact_joint/rod_rigidbody_connection.py

```diff
@@ -0,0 +1,417 @@
+__doc__ = (
+    """ Module containing joint classes to connect rods and rigid bodies together. """
+)
+__all__ = ["RodRigidBodyFreeJoint", "RodRigidBodyFixedJoint"]
+from elastica._rotations import _inv_rotate
+from elastica.joint import FreeJoint, FixedJoint
+from elastica.typing import SystemType
+from elastica.utils import Tolerance, MaxDimension
+import numpy as np
+
+
+class RodRigidBodyFreeJoint(FreeJoint):
+    """
+    This free joint class is the base class for all joints. Free or spherical
+    joints constrains the relative movement between two nodes (chosen by the user)
+    by applying restoring forces. For implementation details, refer to Zhang et al. Nature Communications (2019).
+
+    Notes
+    -----
+    Every new joint class must be derived from the FreeJoint class.
+
+    Attributes
+    ----------
+    k: float
+        Stiffness coefficient of the joint.
+    nu: float
+        Damping coefficient of the joint.
+    point_system_one : numpy.ndarray
+            Describes for system one in the local coordinate system the translation from the node `index_one` (for rods)
+            or the center of mass (for rigid bodies) to the joint.
+    point_system_two : numpy.ndarray
+        Describes for system two in the local coordinate system the translation from the node `index_two` (for rods)
+        or the center of mass (for rigid bodies) to the joint.
```

Comment:
shouldn't one of these refer to the rod and the other to the rigid body now? based on the naming?

Reply:
@bhosale2 I have a different opinion than you on this one. I want to leave the option open for the user to combine systems in any way they would like:

- **Rod to Rod:** In this situation, the users can also use the standard (non-experimental) `FreeJoint` implementation, which should numerically give the same result, but is possibly computationally slightly faster
- **Rod to Rigid Body**
- **Rigid Body to Rod**: This should have the same exact effect as the second bullet point, provided that the user mirrors the configuration.
- **Rigid Body to Rigid Body**: Most importantly, this class should also enable this configuration. If I force the first system to be a rod and the second system to be a rigid body, this is not possible anymore and not implemented anywhere else

I am open though to changing the name of the file / of the classes if you think something would be more intuitive than the current naming scheme.

Reply:
I appreciate the generality seen in the code here, but it is not consistent with the naming of the function then, which when seen one expects a joint between a rod and rigid body. In that case something like a generic_joint would make more sense? generalised_joint? @armantekinalp can voice his opinion as well on this

Reply:
@bhosale2 I see those two classes as a more general implementation of the original FreeJoint and FixedJoint, which should be called something like Rod2RodFreeJoint and Rod2RodFixedJoint in the future. 

So I would be actually in favor in calling the new, more general implementation either FreeJoint and FixedJoint or NewFreeJoint and  NewFixedJoint to communicate to the user that this will be (most likely) the future standard implementation of these joint classes.

Reply:
I agree with the first part, but refactoring the core codebase `FreeJoint` has its own drawbacks. `NewFreejoint` may be a good idea. Anyways, for this PR we are shifting these to experimental as planned, and we can revisit this naming then later in the future.

Reply:
@armantekinalp If you agree as well, I will rename it accordingly to NewFreeJoint and NewFixedJoint.  They are in the experimental module anyway.

Reply:
Hi @mstoelzle . After discussing with @bhosale2 and @skim0119,  we think we can rename them as `GenericSystemTypeFixedJoint` and `GenericSystemTypeFreeJoint`for this PR.

Reply:
Thanks for this suggestion. I renamed in https://github.com/GazzolaLab/PyElastica/pull/149/commits/002770e757697ff071a419358aa75af4d0b9839c